core/src/main/java/org/jdbi/v3/core/qualifier/Qualifiers.java

```diff
@@ -36,40 +38,87 @@
 @Beta
 public class Qualifiers implements JdbiConfig<Qualifiers> {
     private static final JdbiCache<AnnotatedElement[], Set<Annotation>> QUALIFIER_CACHE = JdbiCaches.declare(
-            elements -> elements.length == 1 ? elements[0] : new HashSet<>(Arrays.asList(elements)),
-            (Function<AnnotatedElement[], Set<Annotation>>) Qualifiers::getQualifiers);
+        elements -> elements.length == 1 ? elements[0] : new HashSet<>(Arrays.asList(elements)),
+        (registry, elements) -> registry.get(Qualifiers.class).resolveQualifiers(elements));
+    private final Set<Function<AnnotatedElement, Set<Annotation>>> resolvers = new CopyOnWriteArraySet<>();
     private ConfigRegistry registry;
 
-    public Qualifiers() {}
+    public Qualifiers() {
+        resolvers.add(Qualifiers::getQualifierAnnotations);
+        resolvers.add(Qualifiers::inspectQualifiedCarrier);
+    }
+
+    private Qualifiers(Qualifiers other) {
+        this.resolvers.addAll(other.resolvers);
+        this.registry = null;
```

Comment:
It's a little weird to have an optional registry here.
If you instead do `new ConfigRegistry().get(Qualifiers.class)` for testing does that remove the need for `Qualifiers` to work without the registry?

Reply:
I tried with `new` in the `before` and jdbicache still gave me nullpointers

Reply:
Also, the other block has to be null-safe anyway for the illegitimate uses of this class, so this way at least it's consistent

Reply:
This is what happens with a `new` registry:
![image](https://user-images.githubusercontent.com/3950300/64245176-bd0ed800-cf0a-11e9-9992-90dde9187e34.png)

I'm not quite immediately seeing a logical explanation for it

Reply:
Huh... adding the `new` registry makes `addResolver` not work...

Reply:
The resolver is added to `resolvers` (size = 3), and then in the next invocation on `qualifiers`, it's gone (size = 2)??

Reply:
Ugh, it causes some sort of problem with jdbicache's loader creating a new instance of Qualifiers that doesn't inherit the resolvers list or something :/

Reply:
I overlooked that the solution was in your post all along 😅